cpp/src/arrow/compute/kernels/scalar_compare.cc

```diff
@@ -746,6 +805,199 @@ std::shared_ptr<ScalarFunction> MakeScalarMinMax(std::string name,
+                                       Op<BetweenOptions::Inclusive::BOTH>>::Exec(ctx,
+                                                                                  batch,
+                                                                                  out);
+      case BetweenOptions::Inclusive::LEFT:
+        return ScalarTernaryEqualTypes<BooleanType, FixedSizeBinaryType,
+                                       Op<BetweenOptions::Inclusive::LEFT>>::Exec(ctx,
+                                                                                  batch,
+                                                                                  out);
+      case BetweenOptions::Inclusive::RIGHT:
+        return ScalarTernaryEqualTypes<BooleanType, FixedSizeBinaryType,
+                                       Op<BetweenOptions::Inclusive::RIGHT>>::Exec(ctx,
+                                                                                   batch,
+                                                                                   out);
+      case BetweenOptions::Inclusive::NEITHER:
+        return ScalarTernaryEqualTypes<
+            BooleanType, FixedSizeBinaryType,
+            Op<BetweenOptions::Inclusive::NEITHER>>::Exec(ctx, batch, out);
+      default:
+        return Status::NotImplemented("between inclusiveness not implemented: ",
+                                      state.options.ToString());
+    }
+  } else if (type_id == Type::DECIMAL128 || type_id == Type::DECIMAL256) {
+    switch (state.options.inclusive) {
+      case BetweenOptions::Inclusive::BOTH:
+        return GenerateDecimal<ScalarTernaryEqualTypes, BooleanType,
+                               Op<BetweenOptions::Inclusive::BOTH>>(type_id)(ctx, batch,
+                                                                             out);
+      case BetweenOptions::Inclusive::LEFT:
+        return GenerateDecimal<ScalarTernaryEqualTypes, BooleanType,
+                               Op<BetweenOptions::Inclusive::LEFT>>(type_id)(ctx, batch,
+                                                                             out);
+      case BetweenOptions::Inclusive::RIGHT:
+        return GenerateDecimal<ScalarTernaryEqualTypes, BooleanType,
+                               Op<BetweenOptions::Inclusive::RIGHT>>(type_id)(ctx, batch,
+                                                                              out);
+      case BetweenOptions::Inclusive::NEITHER:
+        return GenerateDecimal<ScalarTernaryEqualTypes, BooleanType,
+                               Op<BetweenOptions::Inclusive::NEITHER>>(type_id)(
+            ctx, batch, out);
+      default:
+        return Status::NotImplemented("between inclusiveness not implemented: ",
+                                      state.options.ToString());
+    }
+  } else if (type_id == Type::BINARY || type_id == Type::STRING ||
+             type_id == Type::LARGE_BINARY || type_id == Type::LARGE_STRING) {
```

Comment:
(untested, but should probably work)
```suggestion
  } else if (is_base_binary_like(type_id)) {
```